pkg/registry/apis/provisioning/jobs/pullrequest/changes.go

```diff
@@ -0,0 +1,235 @@
+	info.Parsed, err = opts.parser.Parse(ctx, fileInfo)
+	if err != nil {
+		info.Error = err.Error()
+		return info, nil
+	}
+
+	// Find a name within the file
+	obj := info.Parsed.Obj
+	info.Title = info.Parsed.Meta.FindTitle(obj.GetName())
+
+	// Check what happens when we apply changes
+	// NOTE: this will also invoke any server side validation
+	err = info.Parsed.DryRun(ctx)
+	if err != nil {
+		info.Error = err.Error()
+		return info, nil
+	}
+
+	// Dashboards get special handling
+	if info.Parsed.GVK.Kind == dashboardKind {
+		info.GrafanaURL = fmt.Sprintf("%sd/%s/%s", baseURL, obj.GetName(), info.Title)
+		info.PreviewURL = baseURL + path.Join("admin/provisioning",
+			info.Parsed.Repo.Name, "dashboard/preview", info.Parsed.Info.Path)
+
+		query := url.Values{}
+		query.Set("ref", info.Parsed.Info.Ref)
+		if opts.pullRequest.URL != "" {
+			query.Set("pull_request_url", url.QueryEscape(opts.pullRequest.URL))
+		}
+		info.PreviewURL += "?" + query.Encode()
+	}
+
+	return info, nil
+}
+
+func calculateFileDeleteInfo(ctx context.Context, baseURL string, change repository.VersionedFileChange, opts changeOptions) (fileChangeInfo, error) {
+	// TODO -- read the old and verify
+	return fileChangeInfo{Change: change, Error: "delete feedback not yet implemented"}, nil
+}
+
+// This will update render the linked screenshots and update the screenshotURLs
+func (f *fileChangeInfo) renderScreenshots(ctx context.Context, baseURL string, renderer ScreenshotRenderer) (err error) {
+	if f.GrafanaURL != "" {
+		f.GrafanaScreenshotURL, err = renderScreenshotFromGrafanaURL(ctx, baseURL, renderer, f.Parsed.Repo, f.GrafanaURL)
+		if err != nil {
+			return err
+		}
+	}
+	if f.PreviewURL != "" {
+		f.PreviewScreenshotURL, err = renderScreenshotFromGrafanaURL(ctx, baseURL, renderer, f.Parsed.Repo, f.PreviewURL)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func renderScreenshotFromGrafanaURL(ctx context.Context,
+	baseURL string,
+	renderer ScreenshotRenderer,
+	repo provisioning.ResourceRepositoryInfo,
+	grafanaURL string,
+) (string, error) {
+	parsed, err := url.Parse(grafanaURL)
+	if err != nil {
+		logging.FromContext(ctx).Warn("invalid", "url", grafanaURL, "err", err)
+		return "", err
+	}
+	snap, err := renderer.RenderScreenshot(ctx, repo, strings.TrimPrefix(parsed.Path, "/"), parsed.Query())
+	if err != nil {
+		logging.FromContext(ctx).Warn("render failed", "url", grafanaURL, "err", err)
+		return "", fmt.Errorf("error rendering screenshot %w", err)
+	}
+	if strings.Contains(snap, "://") {
+		return snap, nil // it is a full URL already (can happen when the blob storage returns CDN urls)
+	}
+	base, err := url.Parse(baseURL)
+	if err != nil {
+		logger.Warn("invalid base", "url", baseURL, "err", err)
+		return "", err
+	}
+	return base.JoinPath(snap).String(), nil
+}
+
+// Remove files we should not try to process
+func onlySupportedFiles(files []repository.VersionedFileChange) (ret []repository.VersionedFileChange) {
```

Comment:
given that this one is only used in worker.go for early returns, I think it's better to put it there.